### PR TITLE
Add DINOv2 with registers (using diff)

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR is a follow-up of #31832 but uses the diff instead of rewriting from scratch.

Had 2 issues with the diff:

- [ ] the `drop_path` function was causing an issue. Applying the suggestion at [#31832](https://github.com/huggingface/transformers/pull/31832#issuecomment-2298687531) fixed it.
- [ ] the diff doesn't properly add `super().__init__()` to the Backbone class. I manually added it in the modeling file as otherwise there's an error.

### CODE REVIEW

File: src/transformers/models/dinov2_with_registers/modeling_dinov2_with_registers.py
Comment (ArthurZucker) [inline src/transformers/models/dinov2_with_registers/modeling_dinov2_with_registers.py:1]: was this actually generate with the diff converted? (I won't review it in this case!)

File: src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        patch_pos_embed = patch_pos_embed.reshape(1, int(math.sqrt(num_positions)), int(math.sqrt(num_positions)), dim)
+        patch_pos_embed = patch_pos_embed.permute(0, 3, 1, 2)
+        target_dtype = patch_pos_embed.dtype
+        patch_pos_embed = nn.functional.interpolate(
+            patch_pos_embed.to(dtype=torch.float32),
+            scale_factor=(float(height / math.sqrt(num_positions)), float(width / math.sqrt(num_positions))),
+            mode="bicubic",
+            align_corners=False,
+            antialias=self.config.interpolate_antialias,
+        ).to(dtype=target_dtype)
```
Comment (ArthurZucker) [inline src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py:235]: can be split!

File: src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+        if self.register_tokens is not None:
+            embeddings = torch.cat(
+                (
+                    embeddings[:, :1],
+                    self.register_tokens.expand(embeddings.shape[0], -1, -1),
+                    embeddings[:, 1:],
+                ),
+                dim=1,
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py:266]: same here to split

File: src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+        # add the [CLS] token to the embedded patch tokens
+        cls_tokens = self.cls_token.expand(batch_size, -1, -1)
+        embeddings = torch.cat((cls_tokens, embeddings), dim=1)
+
+        # add positional encoding to each token
+        embeddings = embeddings + self.interpolate_pos_encoding(embeddings, height, width)
+
+        if self.register_tokens is not None:
```
Comment (ArthurZucker) [inline src/transformers/models/dinov2_with_registers/diff_dinov2_with_registers.py:258]: tokens should always be register as this is register dino2!

**Review by ArthurZucker — COMMENTED** [review]
Super nice! thanks for using diff, really fits well

File: src/transformers/models/dinov2_with_registers/modeling_dinov2_with_registers.py
Comment (NielsRogge) [reply src/transformers/models/dinov2_with_registers/modeling_dinov2_with_registers.py:1]: Yes it was

**Comment by NielsRogge** [comment]
@ArthurZucker thanks for the review, addressed your comments. Note that there were still some issues with the diff converter (see my first message), but other than that this PR seems ready

**Comment by codybum** [comment]
Support for dinov2 with registers would be very much appreciated!

**Comment by ArthurZucker** [comment]
#33248 will resolve all your issues 🤗

**Comment by ArthurZucker** [comment]
I'll ask you to wait a tiny little bit as it will conflict with this work!

**Comment by codybum** [comment]
> @ArthurZucker thanks for the review, addressed your comments. Note that there were still some issues with the diff converter (see my first message), but other than that this PR seems ready

Looks like Modular transformers: modularity and inheritance for new model additions https://github.com/huggingface/transformers/pull/33248 was merged a few hours ago!

**Comment by ArthurZucker** [comment]
Yep! Make sure to rebase!

**Comment by codybum** [comment]
I think I did the wrong @ initially. @NielsRogge do you have time to rebase and update?  If not, we can give it a go. 

> > @ArthurZucker thanks for the review, addressed your comments. Note that there were still some issues with the diff converter (see my first message), but other than that this PR seems ready
> 
> Looks like Modular transformers: modularity and inheritance for new model additions #33248 was merged a few hours ago!

**Comment by NielsRogge** [comment]
Hi, feel free to take over, it's probably easy as only one method needs to be overwritten.

**Comment by NielsRogge** [comment]
Closing this one, will try out the new modular Transformers when I have time

**Pull request closed without merging**